File: advent-of-code/2015/01/02/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

func findBasementPosition(source string) int {
	floor := 0
	for position, char := range source {
		switch char {
		case '(':
			floor++
		case ')':
			floor--
		}

		if floor == -1 {
			return position + 1 // 1-indexed
		}
	}
	return 0
}

func main() {
	// Test cases with assertions
	if findBasementPosition(")") != 1 {
		panic("Test failed: )")
	}
	if findBasementPosition("()())") != 5 {
		panic("Test failed: ()())")
	}

	// Calculate and print the answer
	_, filename, _, _ := runtime.Caller(0)
	scriptDir := filepath.Dir(filename)
	inputPath := filepath.Join(scriptDir, "..", "input.txt")
	data, _ := os.ReadFile(inputPath)
	fmt.Printf("Go: %d\n", findBasementPosition(string(data)))
}

File: advent-of-code/2015/01/02/solution.zig
const std = @import("std");

fn findBasementPosition(source: []const u8) usize {
    var floor: i32 = 0;
    for (source, 0..) |char, index| {
        if (char == '(') {
            floor += 1;
        } else if (char == ')') {
            floor -= 1;
        }

        if (floor == -1) {
            return index + 1; // 1-indexed
        }
    }
    return 0;
}

pub fn main() !void {
    // Test cases with assertions
    std.debug.assert(findBasementPosition(")") == 1);
    std.debug.assert(findBasementPosition("()())") == 5);

    // Calculate and print the answer
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const input = try std.fs.cwd().readFileAlloc(allocator, "../input.txt", 1024 * 1024);
    defer allocator.free(input);

    std.debug.print("Zig: {d}\n", .{findBasementPosition(input)});
}
